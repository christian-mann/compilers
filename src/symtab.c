#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<stdbool.h>
#include "symtab.h"

void initTable(char*);
SymbolTable* createNode(char*);
void printTable(SymbolTable*, int);

SymbolTable *root;
SymbolTable *frontier;
bool shouldTurn;

void initTable(char* word) {
	root = frontier = createNode(word);
	shouldTurn = true;
}

SymbolTable* createNode(char* word) {
	SymbolTable* new = malloc(sizeof(SymbolTable));
	new->entry = malloc(sizeof(SymbolTableEntry));
	new->entry->word = malloc(strlen(word)*sizeof(char));
	strcpy(new->entry->word, word);
	return new;
}

SymbolTableEntry* checkSymbolTable(char* word) {
	if(!frontier) return NULL;

	SymbolTable* pTab = frontier;
	while(true) {
		if(!pTab->entry) {
			return NULL;
		}
		if(!strcmp(word, pTab->entry->word)) {
			return pTab->entry;
		}
		if(pTab->prev) {
			pTab = pTab->prev;
		} else if(pTab->parent) {
			pTab = pTab->parent;
		} else break;
	}
	return NULL;
}

void addVariable(char* word) {
	if(!frontier) fprintf(stderr, "Adding variable before entering procedure!\n");

	SymbolTable* new = createNode(word);

	if(shouldTurn) {
		new->parent = frontier;
		frontier->lchild = new;
		shouldTurn = false;
	} else {
		//add sibling
		new->prev = frontier;
		new->parent = frontier->parent;
		frontier->next = new;
	}
	frontier = new;
}

void enterProcedure(char* word) {
	if(!frontier) {
		initTable(word);
	} else {
		//add sibling
		SymbolTable* new = createNode(word);
		new->prev = frontier;
		new->parent = frontier->parent;
		frontier->next = new;

		frontier = new;
	}
	shouldTurn = true;
}

void exitProcedure() {
	if(!frontier) fprintf(stderr, "Error: Stack Underflow\n");
	//up to parent
	frontier = frontier->parent;

	shouldTurn = false;
}
