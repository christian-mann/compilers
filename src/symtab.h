#ifndef SYMTAB_H
#define SYMTAB_H
typedef struct {
	int type;
	char *word;
} SymbolTableEntry;

typedef struct symtab {
	SymbolTableEntry* entry;
	struct symtab* parent;
	struct symtab* lchild;
	struct symtab* prev;
	struct symtab* next;
} SymbolTable;

SymbolTableEntry* checkSymbolTable(char*);
void addVariable(char*);
void enterProcedure(char*);
void exitProcedure(void);
#endif
