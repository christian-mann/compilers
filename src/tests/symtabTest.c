
#include<stdio.h>
#include "../symtab.h"

#define p enterProcedure
#define v addVariable
#define up exitProcedure
extern SymbolTable *root;

void printTable(SymbolTable*, int);
int main() {
	p("test");
		v("input");
		v("output");
		v("a");
		p("p1");
			v("a");
			v("b");
			v("c");
			p("p2");
				v("d");
			up();
		up();
		p("p3");
			v("x");
		up();
		p("p4");
			v("z");
			v("q");
		up();
	up();

	printTable(root, 0);
}

void printTable(SymbolTable* pTab, int depth) {
	for(int i = 0; i < depth; i++) putchar(' ');
	printf("%s\n", pTab->entry->word);
	if(pTab->lchild) printTable(pTab->lchild, depth + 1);
	if(pTab->next) printTable(pTab->next, depth);
}
