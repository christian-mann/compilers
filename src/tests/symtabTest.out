test
 input
 output
 a
 p1
  a
  b
  c
  p2
   d
 p3
  x
 p4
  z
  q
